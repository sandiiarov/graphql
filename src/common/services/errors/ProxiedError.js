// @flow

type ProxiedErrorValues = {|
  +statusCode: string,
  +url: string,
|};

/**
 * This exception is thrown if origin server does NOT respond with HTTP status code 200.
 */
class ProxiedError extends Error {
  extensions: {|
    +_proxy: ProxiedErrorValues,
    +extensions: {|
      +proxy: ProxiedErrorValues,
    |},
  |};

  constructor(message: string, originStatusCode: string, originUrl: string) {
    super(message);
    this.extensions = {
      _proxy: {
        // note: this is deprecated and will be removed
        statusCode: originStatusCode,
        url: originUrl,
      },
      extensions: {
        proxy: {
          statusCode: String(originStatusCode),
          url: originUrl,
        },
      },
    };
  }
}

export { ProxiedError };
