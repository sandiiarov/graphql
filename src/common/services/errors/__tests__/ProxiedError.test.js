// @flow

import { ProxiedError } from '../ProxiedError';

let error;

beforeEach(() => {
  error = new ProxiedError('custom message', '123', 'http://a.b');
});

it('is throwable', () => {
  expect(() => {
    throw error;
  }).toThrow('custom message');
});

it('adds correct extensions', () => {
  expect(error.extensions).toMatchSnapshot();
});
