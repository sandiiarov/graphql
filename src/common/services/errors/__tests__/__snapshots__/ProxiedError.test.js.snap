// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds correct extensions 1`] = `
Object {
  "_proxy": Object {
    "statusCode": "123",
    "url": "http://a.b",
  },
  "extensions": Object {
    "proxy": Object {
      "statusCode": "123",
      "url": "http://a.b",
    },
  },
}
`;
